File: python/tree234.py
"""Generic array-backed 2-3-4 tree with monoidal measures.

Each node can hold 1, 2, or 3 keys (2-node, 3-node, 4-node).
Nodes are stored in a flat list with a free-list for reuse.
Measures are aggregated bottom-up for efficient weight-based lookups.
"""


class _Node:
    __slots__ = ['n', 'keys', 'children', 'measure']

    def __init__(self):
        self.n = 0
        self.keys = [None, None, None]
        self.children = [-1, -1, -1, -1]
        self.measure = None


class Tree234:
    """Array-backed 2-3-4 tree with monoidal measures."""

    def __init__(self, measure_fn, combine_fn, identity_fn, compare_fn):
        """
        measure_fn: key -> measure (leaf measure for a single key)
        combine_fn: (measure, measure) -> measure (monoidal combine)
        identity_fn: () -> measure (monoidal identity)
        compare_fn: (key, key) -> int (-1, 0, 1)
        """
        self._nodes = []
        self._free = []
        self._root = -1
        self._count = 0
        self._measure_fn = measure_fn
        self._combine = combine_fn
        self._identity = identity_fn
        self._compare = compare_fn

    # -- node allocation ---------------------------------------------------

    def _alloc(self):
        """Allocate a node, reusing from free list if possible."""
        if self._free:
            idx = self._free.pop()
            node = self._nodes[idx]
            node.n = 0
            node.keys = [None, None, None]
            node.children = [-1, -1, -1, -1]
            node.measure = None
            return idx
        idx = len(self._nodes)
        self._nodes.append(_Node())
        return idx

    def _nd(self, idx):
        return self._nodes[idx]

    def _is_leaf(self, idx):
        return self._nodes[idx].children[0] == -1

    # -- measure maintenance -----------------------------------------------

    def _recompute_measure(self, idx):
        """Recompute the aggregate measure for node idx from its children and keys."""
        nd = self._nodes[idx]
        m = self._identity()
        for i in range(nd.n + 1):
            if nd.children[i] != -1:
                m = self._combine(m, self._nodes[nd.children[i]].measure)
            if i < nd.n:
                m = self._combine(m, self._measure_fn(nd.keys[i]))
        nd.measure = m

    # -- public interface --------------------------------------------------

    def size(self):
        """Return the number of keys in the tree."""
        return self._count

    def root_measure(self):
        """Return the aggregate measure of the entire tree."""
        if self._root == -1:
            return self._identity()
        return self._nodes[self._root].measure

    def clear(self):
        """Remove all keys."""
        self._nodes.clear()
        self._free.clear()
        self._root = -1
        self._count = 0

    def insert(self, key):
        """Insert a key using top-down 4-node splitting."""
        if self._root == -1:
            self._root = self._alloc()
            nd = self._nd(self._root)
            nd.n = 1
            nd.keys[0] = key
            nd.measure = self._measure_fn(key)
            self._count = 1
            return

        # If root is a 4-node, split it
        if self._nodes[self._root].n == 3:
            old_root = self._root
            self._root = self._alloc()
            nd = self._nd(self._root)
            nd.n = 0
            nd.children[0] = old_root
            self._split_child(self._root, 0)

        self._insert_topdown(self._root, key)
        self._count += 1

    def to_list(self):
        """In-order traversal returning all keys as a list."""
        result = []
        if self._root != -1:
            self._inorder(self._root, result)
        return result

    def find_by_weight(self, target, weight_fn):
        """Walk down the tree tracking cumulative weight.

        weight_fn: measure -> float (extracts a scalar weight from a measure)

        Returns (key, cumulative_weight_before, in_order_index) or None.
        """
        if self._root == -1:
            return None
        return self._find_by_weight_rec(self._root, target, weight_fn, 0.0, 0)

    def find_neighbors(self, key, weight_fn):
        """Find a key's position and its in-order neighbors.

        weight_fn: measure -> float (extracts scalar weight)

        Returns a dict:
          'found': bool
          'node_key': the actual key object if found, else None
          'left': predecessor key or None
          'right': successor key or None
          'cum_weight': cumulative weight before found/insert position
          'index': in-order index of found key or insertion point
        """
        result = {
            'found': False,
            'node_key': None,
            'left': None,
            'right': None,
            'cum_weight': 0.0,
            'index': 0,
        }
        if self._root != -1:
            self._find_neighbors_rec(self._root, key, weight_fn, result, 0.0, 0)
        return result

    def update_measures_for_key(self, key_obj, weight_fn):
        """Given a key object already in the tree (identity match), recompute
        measures on the path from that key to the root.

        key_obj must be the exact object (is-check) stored in the tree.
        Returns True if found and updated.
        """
        if self._root == -1:
            return False
        return self._update_measures_rec(self._root, key_obj, weight_fn)

    # -- top-down insert internals -----------------------------------------

    def _split_child(self, parent_idx, child_pos):
        """Split a 4-node child at child_pos, pushing middle key to parent."""
        parent = self._nd(parent_idx)
        child_idx = parent.children[child_pos]
        child = self._nd(child_idx)

        mid_key = child.keys[1]

        # Create right sibling with key[2] and children[2..3]
        right_idx = self._alloc()
        right = self._nd(right_idx)
        right.n = 1
        right.keys[0] = child.keys[2]
        right.children[0] = child.children[2]
        right.children[1] = child.children[3]

        # Shrink child to keep only key[0] and children[0..1]
        child.n = 1
        child.keys[1] = None
        child.keys[2] = None
        child.children[2] = -1
        child.children[3] = -1

        # Make room in parent
        for i in range(parent.n, child_pos, -1):
            parent.keys[i] = parent.keys[i - 1]
        for i in range(parent.n + 1, child_pos + 1, -1):
            parent.children[i] = parent.children[i - 1]

        parent.keys[child_pos] = mid_key
        parent.children[child_pos + 1] = right_idx
        parent.n += 1

        self._recompute_measure(child_idx)
        self._recompute_measure(right_idx)
        self._recompute_measure(parent_idx)

    def _insert_topdown(self, idx, key):
        """Recursively insert into a non-full node, splitting 4-node children
        on the way down."""
        nd = self._nd(idx)

        if self._is_leaf(idx):
            # Find insertion position
            pos = nd.n
            for i in range(nd.n):
                if self._compare(key, nd.keys[i]) <= 0:
                    pos = i
                    break
            # Shift keys right
            for i in range(nd.n, pos, -1):
                nd.keys[i] = nd.keys[i - 1]
            nd.keys[pos] = key
            nd.n += 1
            self._recompute_measure(idx)
            return

        # Find child to descend into
        pos = nd.n
        for i in range(nd.n):
            if self._compare(key, nd.keys[i]) <= 0:
                pos = i
                break

        child_idx = nd.children[pos]

        # Pre-split if child is a 4-node
        if self._nodes[child_idx].n == 3:
            self._split_child(idx, pos)
            # Decide which side of the promoted key to go
            if self._compare(key, nd.keys[pos]) > 0:
                pos += 1
            child_idx = nd.children[pos]

        self._insert_topdown(child_idx, key)
        self._recompute_measure(idx)

    # -- traversal ---------------------------------------------------------

    def _inorder(self, idx, result):
        nd = self._nodes[idx]
        for i in range(nd.n):
            if nd.children[i] != -1:
                self._inorder(nd.children[i], result)
            result.append(nd.keys[i])
        if nd.children[nd.n] != -1:
            self._inorder(nd.children[nd.n], result)

    # -- weight-based search -----------------------------------------------

    def _find_by_weight_rec(self, idx, target, weight_fn, cum, io_idx):
        nd = self._nodes[idx]
        for i in range(nd.n):
            # Add left child subtree weight
            if nd.children[i] != -1:
                child_w = weight_fn(self._nodes[nd.children[i]].measure)
                if cum + child_w >= target:
                    return self._find_by_weight_rec(
                        nd.children[i], target, weight_fn, cum, io_idx)
                cum += child_w
                io_idx += self._subtree_size(nd.children[i])
            # Check this key
            key_w = weight_fn(self._measure_fn(nd.keys[i]))
            if cum + key_w >= target:
                return (nd.keys[i], cum, io_idx)
            cum += key_w
            io_idx += 1

        # Last child
        if nd.children[nd.n] != -1:
            return self._find_by_weight_rec(
                nd.children[nd.n], target, weight_fn, cum, io_idx)

        # Edge case: target exceeds total weight; return last key
        return (nd.keys[nd.n - 1], cum - weight_fn(self._measure_fn(nd.keys[nd.n - 1])),
                io_idx - 1)

    def _subtree_size(self, idx):
        """Count total keys in subtree. Uses the measure's count field if
        available, otherwise walks the tree."""
        if idx == -1:
            return 0
        # Try to use the count component of measure
        m = self._nodes[idx].measure
        if hasattr(m, 'count'):
            return m.count
        # Fallback: walk
        nd = self._nodes[idx]
        c = nd.n
        for i in range(nd.n + 1):
            if nd.children[i] != -1:
                c += self._subtree_size(nd.children[i])
        return c

    # -- neighbor finding --------------------------------------------------

    def _find_neighbors_rec(self, idx, key, weight_fn, result, cum, io_idx):
        nd = self._nodes[idx]
        for i in range(nd.n):
            # Left child
            if nd.children[i] != -1:
                child_w = weight_fn(self._nodes[nd.children[i]].measure)
                child_sz = self._subtree_size(nd.children[i])
            else:
                child_w = 0.0
                child_sz = 0

            cmp = self._compare(key, nd.keys[i])

            if cmp < 0:
                # key < nd.keys[i]: answer is in left subtree
                result['right'] = nd.keys[i]
                if nd.children[i] != -1:
                    self._find_neighbors_rec(
                        nd.children[i], key, weight_fn, result, cum, io_idx)
                else:
                    result['cum_weight'] = cum
                    result['index'] = io_idx
                return

            if cmp == 0:
                # Found
                cum += child_w
                io_idx += child_sz
                result['found'] = True
                result['node_key'] = nd.keys[i]
                result['cum_weight'] = cum
                result['index'] = io_idx

                # Left neighbor: rightmost in left subtree or last tracked
                if nd.children[i] != -1:
                    result['left'] = self._rightmost(nd.children[i])

                # Right neighbor: leftmost in right subtree or next key
                right_child = nd.children[i + 1] if i + 1 <= nd.n else -1
                if right_child != -1:
                    result['right'] = self._leftmost(right_child)
                elif i + 1 < nd.n:
                    result['right'] = nd.keys[i + 1]
                return

            # cmp > 0: key > nd.keys[i], continue right
            cum += child_w
            io_idx += child_sz

            result['left'] = nd.keys[i]
            cum += weight_fn(self._measure_fn(nd.keys[i]))
            io_idx += 1

        # Descended past all keys; check last child
        if nd.children[nd.n] != -1:
            self._find_neighbors_rec(
                nd.children[nd.n], key, weight_fn, result, cum, io_idx)
        else:
            result['cum_weight'] = cum
            result['index'] = io_idx

    def _rightmost(self, idx):
        """Return the rightmost key in the subtree rooted at idx."""
        nd = self._nodes[idx]
        if nd.children[nd.n] != -1:
            return self._rightmost(nd.children[nd.n])
        return nd.keys[nd.n - 1]

    def _leftmost(self, idx):
        """Return the leftmost key in the subtree rooted at idx."""
        nd = self._nodes[idx]
        if nd.children[0] != -1:
            return self._leftmost(nd.children[0])
        return nd.keys[0]

    # -- in-place key update -----------------------------------------------

    def _update_measures_rec(self, idx, key_obj, weight_fn):
        """Find key_obj by identity (is-check) and recompute measures upward."""
        nd = self._nodes[idx]
        for i in range(nd.n):
            if nd.keys[i] is key_obj:
                self._recompute_measure(idx)
                return True

            cmp = self._compare(key_obj, nd.keys[i])
            if cmp < 0:
                if nd.children[i] != -1:
                    if self._update_measures_rec(nd.children[i], key_obj, weight_fn):
                        self._recompute_measure(idx)
                        return True
                return False

            # cmp >= 0: might be equal but different object, or greater
            if cmp == 0:
                # Same sort key but different object; could be in left subtree
                if nd.children[i] != -1:
                    if self._update_measures_rec(nd.children[i], key_obj, weight_fn):
                        self._recompute_measure(idx)
                        return True

        # Check last child
        if nd.children[nd.n] != -1:
            if self._update_measures_rec(nd.children[nd.n], key_obj, weight_fn):
                self._recompute_measure(idx)
                return True
        return False

File: go/tree234.go
// Package tdigest provides a generic array-backed 2-3-4 tree with monoidal
// measures. The tree supports top-down insertion (splitting 4-nodes on the
// way down), in-order traversal, weight-based search, and neighbor finding.
package tdigest

// MeasureOps defines the operations needed for a monoidal measure over keys.
type MeasureOps[K any, M any] struct {
	Measure  func(K) M
	Combine  func(M, M) M
	Identity func() M
	Compare  func(K, K) int
}

// node234 is an internal node in the 2-3-4 tree, stored in a flat array.
// n is the number of keys (1, 2, or 3). children[i] == -1 means no child.
type node234[K any, M any] struct {
	n        int
	keys     [3]K
	children [4]int32
	measure  M
}

// Tree234 is a generic array-backed 2-3-4 tree with monoidal measures.
type Tree234[K any, M any] struct {
	nodes    []node234[K, M]
	freeList []int32
	root     int32
	count    int
	ops      MeasureOps[K, M]
}

// NeighborResult holds the result of a neighbor search.
type NeighborResult[K any] struct {
	HasPred bool
	Pred    K
	HasSucc bool
	Succ    K
	Found   bool // whether exact match was found
	Key     K    // the exact match if Found
}

// NewTree234 creates a new empty 2-3-4 tree with the given measure operations.
func NewTree234[K any, M any](ops MeasureOps[K, M]) *Tree234[K, M] {
	return &Tree234[K, M]{
		root: -1,
		ops:  ops,
	}
}

// Size returns the number of keys in the tree.
func (t *Tree234[K, M]) Size() int {
	return t.count
}

// Clear removes all keys from the tree.
func (t *Tree234[K, M]) Clear() {
	t.nodes = t.nodes[:0]
	t.freeList = t.freeList[:0]
	t.root = -1
	t.count = 0
}

// RootMeasure returns the monoidal measure of the entire tree.
func (t *Tree234[K, M]) RootMeasure() M {
	if t.root == -1 {
		return t.ops.Identity()
	}
	return t.nodes[t.root].measure
}

// allocNode allocates a new node, reusing from free list if possible.
func (t *Tree234[K, M]) allocNode() int32 {
	if len(t.freeList) > 0 {
		idx := t.freeList[len(t.freeList)-1]
		t.freeList = t.freeList[:len(t.freeList)-1]
		nd := &t.nodes[idx]
		nd.n = 0
		nd.children = [4]int32{-1, -1, -1, -1}
		nd.measure = t.ops.Identity()
		return idx
	}
	idx := int32(len(t.nodes))
	t.nodes = append(t.nodes, node234[K, M]{
		children: [4]int32{-1, -1, -1, -1},
		measure:  t.ops.Identity(),
	})
	return idx
}

// freeNode returns a node to the free list.
func (t *Tree234[K, M]) freeNode(idx int32) {
	t.freeList = append(t.freeList, idx)
}

// isLeaf returns true if the node has no children.
func (t *Tree234[K, M]) isLeaf(idx int32) bool {
	return t.nodes[idx].children[0] == -1
}

// isFull returns true if the node has 3 keys (a 4-node).
func (t *Tree234[K, M]) isFull(idx int32) bool {
	return t.nodes[idx].n == 3
}

// recomputeMeasure recomputes the monoidal measure for a node from its keys
// and children.
func (t *Tree234[K, M]) recomputeMeasure(idx int32) {
	nd := &t.nodes[idx]
	m := t.ops.Identity()
	for i := 0; i <= nd.n; i++ {
		if nd.children[i] != -1 {
			m = t.ops.Combine(m, t.nodes[nd.children[i]].measure)
		}
		if i < nd.n {
			m = t.ops.Combine(m, t.ops.Measure(nd.keys[i]))
		}
	}
	nd.measure = m
}

// insertKeyChild inserts a key and right child into a non-full node at the
// given position, shifting existing keys/children to the right.
func (t *Tree234[K, M]) insertKeyChild(idx int32, pos int, key K, rightChild int32) {
	nd := &t.nodes[idx]
	// Shift keys and children right
	for i := nd.n; i > pos; i-- {
		nd.keys[i] = nd.keys[i-1]
		nd.children[i+1] = nd.children[i]
	}
	nd.keys[pos] = key
	nd.children[pos+1] = rightChild
	nd.n++
}

// splitChild splits the child at position childPos of parent parentIdx.
// The child must be a 4-node (3 keys). The middle key is promoted to the
// parent, and the child is split into two 1-key nodes.
func (t *Tree234[K, M]) splitChild(parentIdx int32, childPos int) {
	childIdx := t.nodes[parentIdx].children[childPos]
	child := &t.nodes[childIdx]

	// Create new right node with the third key and right children
	newRight := t.allocNode()
	nr := &t.nodes[newRight]
	nr.n = 1
	nr.keys[0] = child.keys[2]
	nr.children[0] = child.children[2]
	nr.children[1] = child.children[3]

	// Middle key to promote
	midKey := child.keys[1]

	// Shrink child to 1-key node (left part)
	child.n = 1
	child.children[2] = -1
	child.children[3] = -1

	// Recompute measures for the two halves
	t.recomputeMeasure(childIdx)
	t.recomputeMeasure(newRight)

	// Insert middle key into parent, with newRight as the right child
	t.insertKeyChild(parentIdx, childPos, midKey, newRight)
	t.recomputeMeasure(parentIdx)
}

// Insert inserts a key into the tree using top-down insertion.
// 4-nodes are split on the way down.
func (t *Tree234[K, M]) Insert(key K) {
	t.count++

	if t.root == -1 {
		t.root = t.allocNode()
		nd := &t.nodes[t.root]
		nd.n = 1
		nd.keys[0] = key
		nd.measure = t.ops.Measure(key)
		return
	}

	// If root is a 4-node, split it first
	if t.isFull(t.root) {
		oldRoot := t.root
		newRoot := t.allocNode()
		t.nodes[newRoot].children[0] = oldRoot
		t.root = newRoot
		t.splitChild(newRoot, 0)
	}

	cur := t.root
	for {
		nd := &t.nodes[cur]

		// Find insertion position
		pos := 0
		for pos < nd.n {
			cmp := t.ops.Compare(key, nd.keys[pos])
			if cmp <= 0 {
				break
			}
			pos++
		}

		if t.isLeaf(cur) {
			// Insert directly into this leaf
			t.insertKeyChild(cur, pos, key, -1)
			t.recomputeMeasure(cur)
			// Update measures up the path (handled by walking back up)
			// Since we do top-down, we update on the way back via
			// the recursive measure recomputation after insertion.
			t.updateMeasuresRoot()
			return
		}

		// Internal node: descend into the appropriate child
		childIdx := nd.children[pos]
		if t.isFull(childIdx) {
			t.splitChild(cur, pos)
			// After split, re-check which child to descend into
			nd = &t.nodes[cur]
			cmp := t.ops.Compare(key, nd.keys[pos])
			if cmp > 0 {
				pos++
			}
			childIdx = nd.children[pos]
		}

		cur = childIdx
	}
}

// updateMeasuresRoot recomputes measures for all nodes from root.
// For simplicity, we do a recursive recomputation from root.
// This is O(n) but happens only once per insert. A path-based approach
// would be O(log n) but requires tracking the path.
func (t *Tree234[K, M]) updateMeasuresRoot() {
	if t.root != -1 {
		t.updateMeasuresRec(t.root)
	}
}

func (t *Tree234[K, M]) updateMeasuresRec(idx int32) {
	nd := &t.nodes[idx]
	for i := 0; i <= nd.n; i++ {
		if nd.children[i] != -1 {
			t.updateMeasuresRec(nd.children[i])
		}
	}
	t.recomputeMeasure(idx)
}

// ToSlice returns all keys in the tree in sorted (in-order) order.
func (t *Tree234[K, M]) ToSlice() []K {
	result := make([]K, 0, t.count)
	if t.root != -1 {
		t.inOrder(t.root, &result)
	}
	return result
}

func (t *Tree234[K, M]) inOrder(idx int32, result *[]K) {
	nd := &t.nodes[idx]
	for i := 0; i <= nd.n; i++ {
		if nd.children[i] != -1 {
			t.inOrder(nd.children[i], result)
		}
		if i < nd.n {
			*result = append(*result, nd.keys[i])
		}
	}
}

// FindByWeight walks the tree using the monoidal measure to find a key by
// cumulative weight. weightOf extracts the weight component from a measure.
// Returns the key, cumulative weight before it, its index in sorted order,
// and whether it was found.
func (t *Tree234[K, M]) FindByWeight(target float64, weightOf func(M) float64) (key K, cumBefore float64, index int, found bool) {
	if t.root == -1 {
		return
	}

	cur := t.root
	cumBefore = 0.0
	index = 0

	for {
		nd := &t.nodes[cur]
		for i := 0; i <= nd.n; i++ {
			// Check left subtree weight
			if nd.children[i] != -1 {
				childWeight := weightOf(t.nodes[nd.children[i]].measure)
				if cumBefore+childWeight >= target {
					// Descend into this child
					cur = nd.children[i]
					goto next
				}
				cumBefore += childWeight
				index += t.subtreeCount(nd.children[i])
			}

			if i < nd.n {
				keyWeight := weightOf(t.ops.Measure(nd.keys[i]))
				if cumBefore+keyWeight >= target {
					key = nd.keys[i]
					found = true
					return
				}
				cumBefore += keyWeight
				index++
			}
		}
		// Should not reach here if target <= total weight
		// Return last key
		return

	next:
		continue
	}
}

// subtreeCount returns the number of keys in the subtree rooted at idx.
func (t *Tree234[K, M]) subtreeCount(idx int32) int {
	if idx == -1 {
		return 0
	}
	nd := &t.nodes[idx]
	count := nd.n
	for i := 0; i <= nd.n; i++ {
		if nd.children[i] != -1 {
			count += t.subtreeCount(nd.children[i])
		}
	}
	return count
}

// countOf extracts the count component from a measure (number of keys in subtree).
func (t *Tree234[K, M]) countOf(idx int32) int {
	if idx == -1 {
		return 0
	}
	nd := &t.nodes[idx]
	c := nd.n
	for i := 0; i <= nd.n; i++ {
		if nd.children[i] != -1 {
			c += t.countOf(nd.children[i])
		}
	}
	return c
}

// FindNeighbors finds the predecessor and successor of a key in the tree.
// If the key exists in the tree, it is returned in the Found/Key fields.
func (t *Tree234[K, M]) FindNeighbors(key K) NeighborResult[K] {
	var result NeighborResult[K]
	if t.root == -1 {
		return result
	}
	t.findNeighborsRec(t.root, key, &result)
	return result
}

func (t *Tree234[K, M]) findNeighborsRec(idx int32, key K, result *NeighborResult[K]) {
	nd := &t.nodes[idx]
	for i := 0; i < nd.n; i++ {
		cmp := t.ops.Compare(key, nd.keys[i])
		if cmp == 0 {
			result.Found = true
			result.Key = nd.keys[i]
			// Predecessor: rightmost key in left subtree, or previous key
			if nd.children[i] != -1 {
				pred := t.rightmost(nd.children[i])
				result.HasPred = true
				result.Pred = pred
			}
			// Successor: leftmost key in right subtree, or next key
			if nd.children[i+1] != -1 {
				succ := t.leftmost(nd.children[i+1])
				result.HasSucc = true
				result.Succ = succ
			} else if i+1 < nd.n {
				result.HasSucc = true
				result.Succ = nd.keys[i+1]
			}
			// If no child-based pred, check previous keys in this node
			if !result.HasPred && i > 0 {
				result.HasPred = true
				result.Pred = nd.keys[i-1]
			}
			return
		} else if cmp < 0 {
			// key < nd.keys[i], so nd.keys[i] is a potential successor
			result.HasSucc = true
			result.Succ = nd.keys[i]
			if nd.children[i] != -1 {
				t.findNeighborsRec(nd.children[i], key, result)
			}
			return
		} else {
			// key > nd.keys[i], so nd.keys[i] is a potential predecessor
			result.HasPred = true
			result.Pred = nd.keys[i]
		}
	}
	// key > all keys in this node, descend into rightmost child
	if nd.children[nd.n] != -1 {
		t.findNeighborsRec(nd.children[nd.n], key, result)
	}
}

// rightmost returns the rightmost (largest) key in the subtree rooted at idx.
func (t *Tree234[K, M]) rightmost(idx int32) K {
	nd := &t.nodes[idx]
	if nd.children[nd.n] != -1 {
		return t.rightmost(nd.children[nd.n])
	}
	return nd.keys[nd.n-1]
}

// leftmost returns the leftmost (smallest) key in the subtree rooted at idx.
func (t *Tree234[K, M]) leftmost(idx int32) K {
	nd := &t.nodes[idx]
	if nd.children[0] != -1 {
		return t.leftmost(nd.children[0])
	}
	return nd.keys[0]
}

// FindByWeightMid is like FindByWeight but uses centroid midpoints:
// it finds the centroid where cumBefore + weight/2 >= target.
// This is useful for quantile interpolation in t-digests.
func (t *Tree234[K, M]) FindByWeightMid(target float64, weightOf func(M) float64, keyWeight func(K) float64) (key K, cumBefore float64, index int, found bool) {
	if t.root == -1 {
		return
	}

	cur := t.root
	cumBefore = 0.0
	index = 0

	for {
		nd := &t.nodes[cur]
		for i := 0; i <= nd.n; i++ {
			if nd.children[i] != -1 {
				childWeight := weightOf(t.nodes[nd.children[i]].measure)
				if cumBefore+childWeight > target {
					cur = nd.children[i]
					goto next2
				}
				cumBefore += childWeight
				index += t.subtreeCount(nd.children[i])
			}

			if i < nd.n {
				kw := keyWeight(nd.keys[i])
				if cumBefore+kw > target {
					key = nd.keys[i]
					found = true
					return
				}
				cumBefore += kw
				index++
			}
		}
		return

	next2:
		continue
	}
}

// CumulativeWeightBefore returns the total weight of all keys that compare
// less than the given key.
func (t *Tree234[K, M]) CumulativeWeightBefore(key K, weightOf func(M) float64, keyWeight func(K) float64) float64 {
	if t.root == -1 {
		return 0
	}
	return t.cumWeightBeforeRec(t.root, key, weightOf, keyWeight)
}

func (t *Tree234[K, M]) cumWeightBeforeRec(idx int32, key K, weightOf func(M) float64, keyWeight func(K) float64) float64 {
	nd := &t.nodes[idx]
	cum := 0.0
	for i := 0; i < nd.n; i++ {
		cmp := t.ops.Compare(key, nd.keys[i])
		if cmp <= 0 {
			// key <= nd.keys[i], descend left
			if nd.children[i] != -1 {
				cum += t.cumWeightBeforeRec(nd.children[i], key, weightOf, keyWeight)
			}
			return cum
		}
		// key > nd.keys[i]: add left child weight + this key weight
		if nd.children[i] != -1 {
			cum += weightOf(t.nodes[nd.children[i]].measure)
		}
		cum += keyWeight(nd.keys[i])
	}
	// key > all keys, add rightmost child
	if nd.children[nd.n] != -1 {
		cum += t.cumWeightBeforeRec(nd.children[nd.n], key, weightOf, keyWeight)
	}
	return cum
}

// FindByMean finds a centroid by its mean value. Returns the centroid,
// cumulative weight before it, its index, and whether it was found.
func (t *Tree234[K, M]) FindByMean(key K, weightOf func(M) float64, keyWeight func(K) float64) (foundKey K, cumBefore float64, index int, found bool) {
	if t.root == -1 {
		return
	}
	return t.findByMeanRec(t.root, key, 0.0, 0, weightOf, keyWeight)
}

func (t *Tree234[K, M]) findByMeanRec(idx int32, key K, cum float64, baseIdx int, weightOf func(M) float64, keyWeight func(K) float64) (foundKey K, cumBefore float64, index int, found bool) {
	nd := &t.nodes[idx]
	for i := 0; i < nd.n; i++ {
		if nd.children[i] != -1 {
			childCount := t.subtreeCount(nd.children[i])
			cmp := t.ops.Compare(key, nd.keys[i])
			if cmp < 0 {
				return t.findByMeanRec(nd.children[i], key, cum, baseIdx, weightOf, keyWeight)
			}
			cum += weightOf(t.nodes[nd.children[i]].measure)
			baseIdx += childCount
		}

		cmp := t.ops.Compare(key, nd.keys[i])
		if cmp == 0 {
			return nd.keys[i], cum, baseIdx, true
		}
		if cmp < 0 {
			return
		}
		cum += keyWeight(nd.keys[i])
		baseIdx++
	}
	if nd.children[nd.n] != -1 {
		return t.findByMeanRec(nd.children[nd.n], key, cum, baseIdx, weightOf, keyWeight)
	}
	return
}

// KeyAtIndex returns the key at the given 0-based sorted index.
func (t *Tree234[K, M]) KeyAtIndex(idx int) (K, bool) {
	if t.root == -1 || idx < 0 || idx >= t.count {
		var zero K
		return zero, false
	}
	return t.keyAtIndexRec(t.root, idx)
}

func (t *Tree234[K, M]) keyAtIndexRec(nodeIdx int32, target int) (K, bool) {
	nd := &t.nodes[nodeIdx]
	for i := 0; i <= nd.n; i++ {
		if nd.children[i] != -1 {
			childSize := t.subtreeCount(nd.children[i])
			if target < childSize {
				return t.keyAtIndexRec(nd.children[i], target)
			}
			target -= childSize
		}
		if i < nd.n {
			if target == 0 {
				return nd.keys[i], true
			}
			target--
		}
	}
	var zero K
	return zero, false
}
